sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs

```diff
@@ -110,39 +126,56 @@ public async override ValueTask<AccessToken> AuthenticateAsync(bool async, Token
             {
                 throw;
             }
+            catch (ProbeRequestResponseException)
+            {
+                // This was an expected response from the IMDS endpoint without the Metadata header set.
+                // Re-issue the request (CreateRequest will add the appropriate header).
+                return await base.AuthenticateAsync(async, context, cancellationToken).ConfigureAwait(false);
+            }
         }
 
-        protected override async ValueTask<AccessToken> HandleResponseAsync(bool async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
+        protected override async ValueTask<AccessToken> HandleResponseAsync(bool async, TokenRequestContext context, HttpMessage message, CancellationToken cancellationToken)
         {
+            Response response = message.Response;
             // if we got a response from IMDS we can stop limiting the network timeout
             _imdsNetworkTimeout = null;
 
             // handle error status codes indicating managed identity is not available
-            var baseMessage = response.Status switch
+            string baseMessage = response.Status switch
             {
+                400 when IsProbRequest(message) => throw new ProbeRequestResponseException(),
                 400 => IdentityUnavailableError,
                 502 => GatewayError,
                 504 => GatewayError,
-                _ => default(string)
+                _ => default
             };
 
             if (baseMessage != null)
             {
-                string message = new RequestFailedException(response, null, new ImdsRequestFailedDetailsParser(baseMessage)).Message;
+                string content = new RequestFailedException(response, null, new ImdsRequestFailedDetailsParser(baseMessage)).Message;
 
                 var errorContentMessage = await GetMessageFromResponse(response, async, cancellationToken).ConfigureAwait(false);
 
                 if (errorContentMessage != null)
                 {
-                    message = message + Environment.NewLine + errorContentMessage;
+                    content = content + Environment.NewLine + errorContentMessage;
                 }
 
-                throw new CredentialUnavailableException(message);
+                throw new CredentialUnavailableException(content);
             }
 
-            return await base.HandleResponseAsync(async, context, response, cancellationToken).ConfigureAwait(false);
+            var token = await base.HandleResponseAsync(async, context, message, cancellationToken).ConfigureAwait(false);
+            return token;
         }
 
+        public static bool IsProbRequest(HttpMessage message)
+            => message.Request.Uri.Host == s_imdsEndpoint.Host &&
```

Comment:
And this holds true even in other clouds or even Stack? Or is it not feasible on Stack anyway?

Reply:
IMDS is always the same well known URI